### ISSUE TITLE
Errors when adding a Number (integer) field to a Layout

### ISSUE THREAD

**Issue by philsward** [backdrop/backdrop-issues#4732]
**Description of the bug**

When adding a Number (integer) field to a layout, the layout page produces errors:

**Steps To Reproduce**

To reproduce the behavior:
1. Add a Number (integer) field to a content type
2. Add the Number (integer) field as a block to a layout
3. Upon save, the error shows up.

**Actual behavior**

List of errors
```
Notice: Undefined index: scale in number_field_formatter_settings_summary() (line 281 of /home/user/public_html/t/sandbox/backdrop1.sandbox/core/modules/field/modules/number/number.module).
Notice: Undefined index: decimal_separator in number_field_formatter_settings_summary() (line 281 of /home/user/public_html/t/sandbox/backdrop1.sandbox/core/modules/field/modules/number/number.module).
```

**Expected behavior**

There should not be any errors.

**Additional information**

Add any other information that could help, such as:
- Backdrop CMS version: 1.17.1
- Web server and its version: Apache 2.4 | PHP 7.2 & 7.3 & 7.4
- Operating System and its version: CentOS 7
- Browser(s) and their versions: Chrome 86.0.4240.111

**Comment by klonos** [comment]
Hmm 🤔 ...these errors seem like https://www.drupal.org/project/drupal/issues/1002734 but that issue was fixed for both D7 and D8 9 years ago (before we fork Backdrop). I have double-checked the patch in that d.org issue, and confirmed that the same code is still in https://github.com/backdrop/backdrop/blob/1.x/core/modules/field/modules/number/number.module#L227

Odd. Perhaps something specific to layouts is tripping it.

**Comment by philsward** [comment]
I found these errors on a sandbox I'm doing some testing on. I have a live
site running BD so I'll test it over there as well when I get a chance.

The sandbox was fresh from the latest version, the live has been upgraded
from 1.16 -> 1.17

**Comment by klonos** [comment]
Yeah, I wanted to quickly test to see if I can reproduce it, but our demo sandboxes are busted at the moment. I'll set up and test locally when I get a chance.

**Comment by klonos** [comment]
...OK, I've figured it out. This has to do with the field formatter summary for numbers. I'll fix a couple of other inconsistencies and file a PR soon(ish).

**Comment by philsward** [comment]
Perfect! Just glad I could help.

**Comment by klonos** [comment]
PR up for review/testing: https://github.com/backdrop/backdrop/pull/3386

@philsward in the meantime, just know that you can still safely use integer field blocks in your layouts. These errors are related to the "cosmetic" block summary, shown only to admins when editing layouts 😉

**Comment by philsward** [comment]
Nice! Thanks for looking into it so quickly. I'll update my end and give it
another test.

**Comment by philsward** [comment]
Tested and looks good. No more errors!

**Comment by indigoxela** [comment]
I confirmed the Notices when adding an integer field to a layout. And I can also confirm that the PR fixes it (tested locally).

The code is looking good, too. RTBC

**Comment by BWPanda** [comment]
Before this is committed, I'd like to know why this error happens in Layouts and not on the Manage Display page: `/admin/structure/types/manage/page/display/default`...?

If I do a `dpm()` of the `$settings` variable, in Layouts it only has two values (hence the errors about the two missing values), but on the Manage Displays page it has 4 values (and therefore no error). Whatever happens in Manage Displays to avoid this, can't we do that in Layouts too?

**Comment by klonos** [comment]
Thanks @BWPanda ...what you say makes sense. I'll take another look as soon as possible.

**Comment by klonos** [comment]
...another interesting thing: both float and decimal field types have `number_decimal` as their display type.

**Comment by klonos** [comment]
> I'd like to know why this error happens in Layouts and not on the Manage Display page ...If I do a `dpm()` of the `$settings` variable, in Layouts it only has two values (hence the errors about the two missing values), but on the Manage Displays page it has 4 values (and therefore no error).

Found it: https://github.com/backdrop/backdrop/blob/1.x/core/modules/field/modules/number/number.module#L198 ...so what happens basically is that both `decimal_separator` and `scale` are being added in `number_field_formatter_info()` for integer fields:

```php
        // The 'decimal_separator' and 'scale' settings are not configurable
        // through the UI, and will therefore keep their default values. They
        // are only present so that the 'number_integer' and 'number_decimal'
        // formatters can use the same code.
        'decimal_separator' => '.',
        'scale' => 0,
```

This is similar to what I did to fix our issue here, only difference is that I have set the decimal separator for integers to blank (`''`) instead of a dot. I believe that blank or `NULL` makes sense for that when it comes to integers ...but then again perhaps it doesn't matter, as decimals are never being added to integers when rendered via php's `number_format()` 🤷

**Comment by klonos** [comment]
@BWPanda here's an alternative approach: https://github.com/backdrop/backdrop/pull/3390

The issue is fixed for newly-placed blocks, but existing integer field blocks will need to be re-saved. So an update hook will be required if we decide that this is a better way.

**Comment by indigoxela** [comment]
Maybe - for consistency - use the same separator `'decimal_separator' => '.',` as in number module code? It doesn't matter anyway, as the scale of `0` will suppress it anyway, as you wrote.

The alternate PR works fine, too. I'll leave it to @BWPanda to decide on the better approach. (Fix it in number_field_formatter_settings_form() or in number_field_formatter_settings_summary().)

Regarding the update hook... good point. Is it worth it then, anyway? Only the settings_summary is currently showing the Notices and the first PR fixed that.

**Comment by BWPanda** [comment]
Sorry, but this still doesn't answer my question: why doesn't the addition of a number field in Layouts use the same initialisation code as other places, where the default values set in `number_field_formatter_info()` are used to avoid this problem automatically?

**Comment by klonos** [comment]
> Maybe - for consistency - use the same separator `'decimal_separator' => '.',` as in number module code?

TBH @indigoxela, I'd rather we changed the other code to replace the dot with blank. It makes more sense to me, since integers do not have decimals nor decimal points.

> It doesn't matter anyway ...

Yup, it doesn't 😅 

> ... why doesn't the addition of a number field in Layouts use the same initialisation code as other places?

Not sure how to further explain things @BWPanda 🤷 ...what I've noticed is that the node display form uses actual displays, whereas for field blocks we use "pseudo instances" (see [here](https://github.com/backdrop/backdrop/blob/1.x/core/modules/field/field.block.inc#L293) and [here](https://github.com/backdrop/backdrop/blob/1.x/core/modules/field/field.block.inc#L44)). These pseudo instances have a display name of `_custom`, and that in turn does not have all the default formatter settings for integer fields:

![image](https://user-images.githubusercontent.com/2423362/98422156-a24a6d00-2093-11eb-9102-58857bcef9b2.png)

After many many hours, I think I now know where to look 😓 ...about to file yet another alternative PR 😅 (please don't be mad at me).

**Comment by klonos** [comment]
...here's another alternative: https://github.com/backdrop/backdrop/pull/3391

The key change is basically this:

```php
$pseudo_instance = $this->createPseudoInstance($this->settings['formatter'], $this->settings['formatter_settings']);
```

to this:

```php
$pseudo_instance = $this->createPseudoInstance($formatter_name, $formatter_settings);
```

...which basically changes the second parameter of `createPseudoInstance()` from `$this->settings['formatter_settings']` (which is missing `decimal_separator` and `scale` for integer field blocks) to `$formatter_settings` (which is basically the actual settings for the specific formatter, including `decimal_separator` and `scale`).

I hope that this is not too confusing 😅

**Comment by BWPanda** [comment]
I think this last one is the best solution as it means any changes to the default settings for fields will be used here too without duplicating code.

**Comment by BWPanda** [comment]
@klonos I recommend closing the other PRs in favour of the latest one to avoid confusion for people wanting to test/review this.

**Comment by klonos** [comment]
I mostly agree @BWPanda, so I have closed https://github.com/backdrop/backdrop/pull/3386 (the first PR), but I think that there is merit in keeping https://github.com/backdrop/backdrop/pull/3390. I actually think that we might need to merge the changes from both those PRs in the final solution. Here's why:

1. start a fresh installation of Backdrop
2. pick one of the OOTB content types (page/post) and add the following fields:
3. integer (leave all settings as default)
4. float (leave all settings as default)
5. decimal (leave all settings as default)
6. create a new "Node layout" layout, with its path set to `node/%`
7. add all 3 fields as blocks in the newly-created layout and save

Head over to `admin/config/development/configuration/single/export`, or manually check the following config files in the `files/config_bla-bla-bla/active` directory:

- Config group: "Field instances" -> Config name: "Node - Page - Decimal" / (or the `field.instance.node.page.field_decimal.json` config file)
- Config group: "Field instances" -> Config name: "Node - Page - Float" / (or the `field.instance.node.page.field_float.json` config file)
- Config group: "Field instances" -> Config name: "Node - Page - Integer" / (or the `field.instance.node.page.field_integer.json` config file)

...notice how all these configurations have a `"display":` section, where all the formatter settings (including `"decimal_separator"` and `"scale"`) exist, for all 3 field instances (including the integer field).

```json
    ...
    "display": {
        "default": {
            "label": "hidden",
            "type": "number_integer",
            "weight": "2",
            "settings": {
                "thousand_separator": "",
                "decimal_separator": ".",  <--- here
                "scale": 0,                <--- here
                "prefix_suffix": true
            },
            "module": "number"
        }
    }
    ...
```

Next inspect the "Node layout" layout configuration:

- Config group: "Layouts" -> Config name: "Node layout" / (or the `layout.layout.node_layout.json` config file)

...scroll down to find the respective block instances, and notice how the decimal and float field blocks have `"decimal_separator"` and `"scale"` specified, whereas these are missing from the integer field block instance:

```json
        "some-random-UUID-here": {
            "plugin": "field:field_block:node-field_float",
            "data": {
                "module": "field",
                "delta": "field_block",
                "settings": {
                    "title_display": "default",
                    "title": "",
                    "style": "default",
                    "block_settings": [],
                    "contexts": [],
                    "label": "hidden",
                    "formatter": "number_decimal",
                    "formatter_settings": {
                        "thousand_separator": "",
                        "decimal_separator": ".",  <--- here
                        "scale": "3",              <--- here
                        "prefix_suffix": 1
                    },
                    ...
                },

            }
        },
        "some-random-UUID-here": {
            "plugin": "field:field_block:node-field_decimal",
            "data": {
                "module": "field",
                "delta": "field_block",
                "settings": {
                    "title_display": "default",
                    "title": "",
                    "style": "default",
                    "block_settings": [],
                    "contexts": [],
                    "label": "hidden",
                    "formatter": "number_decimal",
                    "formatter_settings": {
                        "thousand_separator": ",",
                        "decimal_separator": ".",  <--- here
                        "scale": "2",              <--- here
                        "prefix_suffix": 1
                    },
                    ...
                },
                ...
            }
        },
        "some-random-UUID-here": {
            "plugin": "field:field_block:node-field_integer",
            "data": {
                "module": "field",
                "delta": "field_block",
                "settings": {
                    "title_display": "default",
                    "title": "",
                    "style": "default",
                    "block_settings": [],
                    "contexts": [],
                    "label": "hidden",
                    "formatter": "number_integer",
                    "formatter_settings": {
                        "thousand_separator": " ",  <--- ???
                        "prefix_suffix": 1
                    },
                    ...
                },
                ...
            }
        },
```

If you repeat the same steps in the sandbox for the second PR, you will see that there is consistency in the config files for all 3 field block instances (all 3 have `"decimal_separator"` and `"scale"` specified). I think that we should have consistency between formatter settings and how they are saved in config for both field instances in the various node type displays, as well as for the layout field block instances. The second PR does that exactly, by adding "hidden" `'#type' => 'value'` fields in the block configuration form for integer field blocks. That results in `"decimal_separator"` and `"scale"` being saved in config for the integer field block instance.

Thoughts?

**Comment by klonos** [comment]
>> Maybe - for consistency - use the same separator `'decimal_separator' => '.',` as in number module code?
>
> TBH @indigoxela, I'd rather we changed the other code to replace the dot with blank. It makes more sense to me, since integers do not have decimals nor decimal points.

...I've updated the 2nd PR with a commit that makes it so that the default values for the `$element['decimal_separator']` and `$element['scale']` form fields are pulled from the default settings for the integer formatter: [`24076ea` (#3390)](https://github.com/backdrop/backdrop/pull/3390/commits/24076ea4dd6f3647eabd5ab65d8af26c61207359)

**Comment by philsward** [comment]
Thought: Sorry for opening a can of worms y'all. Glad there's folks smarter
than me who can figure it out though ;-)

**Comment by klonos** [comment]
Thanks for the can of worms @philsward ...we enjoy the occasional challenge, so it's been fun/pleasure 😅

**Comment by BWPanda** [comment]
IMO this issue is fixed (nicely) by the third PR. Coupled with the fact that the second PR adds new (hidden) form values, I think a separate issue is needed for further discussion about that.

**Comment by klonos** [comment]
Fair enough @BWPanda ...I've closed the 2nd PR, and will file a separate issue for that once this is merged.

**Comment by indigoxela** [comment]
@klonos many thanks for getting to the bottom of the matter. :+1: 

I agree that the third approach (fixing it in getAdminPreview()) is the best solution. RTBC

**Comment by BWPanda** [comment]
It'd be nice, if possible, to get another core committer's eyes on this too, just because of the back-and-forth re. which method's better...

**Comment by klonos** [comment]
We are planning for a bug fix release this weekend, so I'll get @quicksketch to have a look 😉 ...but also pinging @herbdool in case he has some time.

**Comment by quicksketch** [comment]
The PR at https://github.com/backdrop/backdrop/pull/3391 looks perfect to me. A straight-forward solution IMO. Thanks @klonos for all the different approaches and compromise. Thanks @BWPanda and @indigoxela for the reviews. And big thank you to @philsward for reporting the issue.
